community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/ASTNode.scala

```diff
@@ -47,10 +47,11 @@ trait ASTNode
       val constructor = this.copyConstructor
       val params = constructor.getParameterTypes
       val args = children.toVector
-      if ((params.length == args.length + 1) && params.last.isAssignableFrom(classOf[InputPosition]))
-        constructor.invoke(this, args :+ this.position: _*).asInstanceOf[this.type]
-      else
-        constructor.invoke(this, args: _*).asInstanceOf[this.type]
+      val hasExtraParam = params.length == args.length + 1
+      val lastParamIsPos = params.last.isAssignableFrom(classOf[InputPosition])
+      val ctorArgs = if (hasExtraParam && lastParamIsPos) args :+ this.position else args
+      val duped = constructor.invoke(this, ctorArgs: _*)
+      duped.asInstanceOf[this.type]
```

Comment:
I know this isn't really changed but you usually bemoan the usage of `this`, maybe we should put a `self =>` in here?